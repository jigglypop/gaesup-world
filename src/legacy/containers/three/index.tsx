"use client";

import MiniMap from "@components/minimap";
import { START_POSITION } from "@constants/index";
import Foooter from "@containers/footer";
import Header from "@containers/header";
import Moving from "@containers/moving";
import { Canvas } from "@react-three/fiber";
import { Physics } from "@react-three/rapier";
import { initBlockOption, initClickerOption } from "@store/options";
import { rotationAtom } from "@store/rotation/atom";
import { zoomAtom } from "@store/zoom/atom";
import { ZoomCamera } from "@store/zoom/zoomCamera";
import { GaesupWorld } from "gaesup-world";
import { useAtom } from "jotai";
import { ReactNode, useMemo } from "react";
import { useLocation } from "react-router-dom";
import * as S from "./styles.css";

export const keyBoardMap = [{ name: "keyD", keys: ["KeyD"], label: "인사" }];

export default function ThreeContainer({
  children,
  isUpdate = false,
}: {
  children: ReactNode;
  isUpdate: boolean;
}) {
  const [{ zoom }] = useAtom(zoomAtom);
  const [rotate] = useAtom(rotationAtom);
  const location = useLocation();

  const HeaderMemo = useMemo(() => <Header />, [location, location.pathname]);
  const FoooterMemo = useMemo(() => <Foooter />, [location, location.pathname]);

  const { clickerOption } = initClickerOption();
  const { blockOption } = initBlockOption();
  const cameraOption = useMemo(() => {
    return {
      YDistance: rotate.y,
      ZDistance: rotate.z * Math.cos(rotate.angle + Math.PI / 4),
      XDistance: rotate.x * Math.sin(rotate.angle + Math.PI / 4),
    };
  }, [rotate]);

  return (
    <GaesupWorld
      urls={{
        characterUrl: "gltf/ally_body.glb",
      }}
      mode={{
        type: "character",
        controller: "clicker",
        control: "normal",
        isButton: true,
      }}
      debug={false}
      cameraOption={cameraOption}
      startPosition={START_POSITION}
      clickerOption={clickerOption}
      block={blockOption}>
      {HeaderMemo}
      {FoooterMemo}
      {location.pathname === "/aggjack/" && <MiniMap />}

      {isUpdate && <Moving />}
      <Canvas
        flat
        shadows
        frameloop={"demand"}
        camera={{
          fov: 50,
          zoom: zoom,
        }}
        className={S.canvas}>
        <ZoomCamera />
        <Physics updateLoop="independent">{children}</Physics>
      </Canvas>
    </GaesupWorld>
  );
}
