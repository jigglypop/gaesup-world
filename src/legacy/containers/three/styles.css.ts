import { flex } from "@styles/recipe/index.css";
import { style } from "@vanilla-extract/css";

export const canvas = style([
  {
    position: "fixed",
    width: "100vw",
    height: "100vh",
    top: 0,
    left: 0,
  },
]);

export const overlay = style([
  flex({
    column: "center",
  }),
  {
    position: "fixed",
    bottom: "0",
    width: "100%",
    height: "10rem",
    zIndex: 10,
  },
]);
