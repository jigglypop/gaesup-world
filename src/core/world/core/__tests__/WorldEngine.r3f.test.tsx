import React from 'react';
import ReactThreeTestRenderer from '@react-three/test-renderer';
import { WorldEngine } from '../WorldEngine';
import * as THREE from 'three';

describe('WorldEngine R3F Tests', () => {
  it('should test raycast without creating new objects', async () => {
    // R3F 컴포넌트로 테스트
    const TestComponent = () => {
      const engine = React.useRef(new WorldEngine());
      
      React.useEffect(() => {
        const origin = new THREE.Vector3(0, 0, 0);
        const direction = new THREE.Vector3(0, 0, 1);
        
        // 여러 번 raycast 호출
        for (let i = 0; i < 100; i++) {
          engine.current.raycast(origin, direction, 10);
        }
      }, []);
      
      return (
        <mesh>
          <boxGeometry args={[1, 1, 1]} />
          <meshBasicMaterial color="red" />
        </mesh>
      );
    };
    
    const renderer = await ReactThreeTestRenderer.create(<TestComponent />);
    
    // Scene graph 확인
    const graph = renderer.toGraph();
    expect(graph).toMatchSnapshot();
    
    // 메모리 누수 확인 - 재사용된 raycaster 인스턴스 확인
    const mesh = renderer.scene.children[0];
    expect(mesh).toBeDefined();
    
    renderer.unmount();
  });
  
  it('should handle objects in scene correctly', async () => {
    const engine = new WorldEngine();
    
    const TestScene = () => {
      return (
        <>
          <mesh name="object1" position={[0, 0, 0]}>
            <boxGeometry />
            <meshBasicMaterial />
          </mesh>
          <mesh name="object2" position={[5, 0, 0]}>
            <sphereGeometry />
            <meshBasicMaterial />
          </mesh>
        </>
      );
    };
    
    const renderer = await ReactThreeTestRenderer.create(<TestScene />);
    
    // Scene 구조 확인
    const meshes = renderer.scene.children.filter(child => child instanceof THREE.Mesh);
    expect(meshes).toHaveLength(2);
    
    // 위치 확인
    expect(meshes[0].position.x).toBe(0);
    expect(meshes[1].position.x).toBe(5);
    
    // findByProps 사용
    const object1 = renderer.scene.findByProps({ name: 'object1' });
    expect(object1).toBeDefined();
    
    // fireEvent 사용 (포인터 이벤트)
    await renderer.fireEvent(object1, 'click');
    
    renderer.unmount();
  });
}); 