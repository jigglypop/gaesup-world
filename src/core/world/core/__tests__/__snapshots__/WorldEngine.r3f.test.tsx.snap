// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`WorldEngine R3F Tests should test raycast without creating new objects 1`] = `
[
  {
    "children": [
      {
        "children": [],
        "name": "",
        "type": "BoxGeometry",
      },
      {
        "children": [],
        "name": "",
        "type": "MeshBasicMaterial",
      },
    ],
    "name": "",
    "type": "Mesh",
  },
]
`;
